src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,761 @@
+		goto out;
+	}
+
+	do {
+		nr = readdir_chunk;
+		rc = dfs_readdirplus(hdl->dfs, obj, &anchor, &nr, dentries, stats);
+		if (rc) {
+			D_ERROR("Readdirplus of '%s' failed: %s (rc=%d)", path, strerror(rc), rc);
+			goto out;
+		}
+
+		for (uint32_t i = 0; i < nr; ++i) {
+			const struct stat *st    = &stats[i];
+			PyObject          *dname = PyUnicode_FromString(dentries[i].d_name);
+			if (dname == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			if (S_ISDIR(st->st_mode)) {
+				rc = PyList_Append(dirs, dname);
+				Py_DECREF(dname); /* PyList_Append does not steal the reference */
+				if (rc) {
+					goto out;
+				}
+				continue;
+			}
+
+			PyObject *item = PyTuple_New(2);
+			if (item == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			PyTuple_SetItem(item, 0, dname); /* steals the reference */
+			PyTuple_SetItem(item, 1, PyLong_FromLong(st->st_size));
+
+			rc = PyList_Append(files, item);
+			Py_DECREF(item); /* PyList_Append does not steal the reference */
+			if (rc < 0) {
+				rc = EIO;
+				goto out;
+			}
+		}
+	} while (nr > 0);
+
+out:
+	D_FREE(dentries);
+	D_FREE(stats);
+
+	if (obj) {
+		dfs_release(obj);
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__torch_read(PyObject *self, PyObject *args)
+{
+	ssize_t            rc     = 0;
+	struct dfs_handle *hdl    = NULL;
+	char              *path   = NULL;
+	dfs_obj_t         *obj    = NULL;
+	PyObject          *buffer = NULL;
+	Py_buffer          bview;
+	d_iov_t            iov;
+	daos_size_t        read = 0;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsO", &hdl, &path, &buffer);
+	assert(hdl->dfs != NULL);
+
+	if (!PyObject_CheckBuffer(buffer)) {
+		PyErr_SetString(PyExc_TypeError,
+				"Expected an object that supports the buffer protocol");
+		return NULL;
+	}
+
+	if (PyObject_GetBuffer(buffer, &bview, PyBUF_WRITE) == -1) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not writable");
+		return NULL;
+	}
+
+	/*
+	  Since python can use buffer like objects that might not have contiguous memory layout,
+	  let's put a guardrail accepting only buffers with contiguous memory region
+	*/
+	if (!PyBuffer_IsContiguous(&bview, 'C')) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not contiguous");
+		PyBuffer_Release(&bview);
+		return NULL;
+	}
+
+	read = bview.len;
+	d_iov_set(&iov, bview.buf, read);
+
+	d_sg_list_t sgl = {
+	    .sg_nr     = 1,
+	    .sg_nr_out = 0,
+	    .sg_iovs   = &iov,
+	};
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup '%s': %s (rc=%ld)", path, strerror(rc), rc);
+		rc = -rc;
+		goto out;
+	}
+
+	rc = dfs_read(hdl->dfs, obj, &sgl, 0 /* offset */, &read, NULL);
+	if (rc) {
+		rc = -rc;
+		goto out;
+	}
+	rc = read;
+	if (read != bview.len) {
+		rc = -EIO;
+	}
+
+out:
+	PyBuffer_Release(&bview);
+
+	if (obj) {
+		int rc2 = dfs_release(obj);
+		if (rc2) {
+			D_ERROR("Could not release object '%s': %s (rc=%d)", path, strerror(rc2),
+				rc2);
```

Comment:
you have to assign rc to rc2 if rc == 0.
same comment applies to other functions of the same use case when releasing the object.